tutorials/desktop/firefox-for-web-developers/firefox-for-web-developers.md

```diff
@@ -0,0 +1,309 @@
+
+The **Toggle Tools** entry has a shortcut of **Ctrl-Shift-I**. Go on, give it a try. Each time you press it, the panel will be opened or closed, depending on its previous state. For historical reasons there's also another shortcut that will do the same thing, and which is often easier to use if you have a full-sized keyboard: the **F12** key. Pressing **F12** then selecting a particular tool with the mouse is usually faster than navigating directly to it using the **Web Developer** menu, hence why most of the entries in that menu are rarely used.
+
+
+### Using the context menu ###
+There is a third way to open the developer tools. This is not a toggle, as it always switches to a particular tab, but is a very common approach when you want to take a closer look at a specific element on the web page. If you right-click on something in the page to open the context menu, you'll find an entry labelled **Inspect Element** towards the bottom. You can select this to open the developer tools with the **Inspector** tab active and the element you clicked on already highlighted within the panel. Let's give it a try:
+
+1. Right click on the "Do more with Ubuntu!" heading near the top of the web page.
+2. Select **Inspect Element** from the context menu.
+3. The developer tools should open with the Inspector tab selected and the `<h1>` element highlighted.
+
+![Inspecting the page heading](images/inspector.png)
+
+As you can see, this method cuts right through the layers of HTML content to highlight the specific element that you right-clicked on, making this a very convenient way to open the tools if you need to work on a particular part of the page.
+
+
+## The inspector##
+Duration: 8:00
+
+All the tabs in the developer tools panel have their uses, but for beginners to web development — or anyone who is simply curious about how a web page works — you only really need to use the **Inspector** and the **Console**. This section and the next will describe them in a little more detail.
+
+The inspector lets you look at the details of the web page's "Document Object Model" (DOM). This is the browser's internal version of the page structure, and the inspector exposes it in a way that makes it look similar to the HTML that was used to create the page. Be clear, though, that although the DOM is initially created from the HTML, it's not quite the same thing. The DOM is a live representation of the page structure, so if the page content has been dynamically modified by some JavaScript code the changes will be reflected in the DOM, and hence in the inspector. The live nature of the DOM also means that you can change things in the inspector and see the results in real-time on the page — ideal for interactively tweaking the design of your page.
```

Comment:
Could you add a link to '"Document Object Model" (DOM)' for anyone who wants to read up more about the DOM? Maybe to https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model ?

Reply:
The tutorial guidelines are strongly against including links in the body of the content that could send users off to another site. I could add a link on the last page, but there's already a general link there to MDN so it's arguably redundant. That MDN page is also a great reference for DOM interfaces, but a really poor one for explaining the DOM further to someone who doesn't already understand it.

Reply:
Oh really? Okay then. I'll go through and remove my commits. I'm not a fan of that guideline though, I might challenge it.

Reply:
Okay I've removed my commits to add links.

Reply:
I wasn't convinced by that guideline at first, but I've come around to that way of thinking. Sending people off to other sites is likely to break the flow of a tutorial; better to let them complete it, then provide some options as to where to go next.

Reply:
Not that this is the place for the discussion, and right now has no bearing on this PR, but I'm going to rant anyway to get my thoughts down, please just ignore me:

People have been making the same argument links "distracting" users since the dawn of the internet about articles etc., and a tutorial is no different. We can see from stats that the majority of people don't make it to the end of a tutorial, just like an article.

I think links are really a fundamental and expected part of the Internet experience, if they lead people away it's because people are using the Internet how it's intended to be used. If someone gets halfway down your tutorial and realises that what they really need to know at this point is what HTML is before they can begin to understand the DOM or the Inspector, then they should be able to make that call.

The ability to do simple inline referencing is one of the real strengths of the web. You no longer need to expect the reader to simply trust you, you can provide links to more authoritative sources right there inline to show that your point has a basis.

I'm all for tweaking the visual treatment of links to make them less distracting for users, or even possibly switching to the `DOM [1]` referencing style like Wikipedia, I just really don't like abdicating the opportunity to provide references in-place.

Reply:
Broadly I agree, and would hate to see inline links removed in general as they are fundamental to the way the web works. But in a tutorial - which is essentially a series of steps where each one builds on the one before - breaking the flow can be more detrimental than is the case on a 'normal' web page. Less intrusive links, such as the `[1]` style, could be a reasonable compromise though.

I did try a different compromise: a list of links on the last page, but some inline links directing readers straight to the last page if they wanted to jump there. Unfortunately I couldn't find a relative path syntax that worked with the markdown format in a reliable* way. Being able to trivially link between steps in a tutorial would make it easier to collect all the links in one place and let the user jump there as needed. It could also be used to include a glossary page, or to make a tutorial more wizard like (e.g. "click here to jump to step 3 if you already have the software installed").

Reply:
Guidelines co-author here! In a nutshell, it's not about a fear of losing or distracting people, it's about the nature of what you are offering and how self-contained and clear it is. I like to think of it as a friend taking 30 min to teach you how to do something, or walking you through a concept.

To put this experience in writing form, one of the most important parts is to state the objectives and prerequisites of the tutorial. Readers need to know what they will get out of it and what they need to have, or need to know, to get to the end. If this is clearly defined, as a reader, you won't need to follow links inside the tutorial to get to the end, because you have been provided with everything you need at the start.

As a teacher, if you need to add contextual information through links, feel free to do so at the beginning or the end, but your tutorial should be self-contained enough to take people to the end without needing external resources (unless an external action is required at some point of the tutorial).

Of course, this way of sharing knowledge doesn't fit every content, and in general, I think we aim at offering two types of tutorials:

  * 10m action: for example, install or configure a piece of software (less than ~30 minutes)
  * 1h project: put several pieces together, maybe write some code and get something tangible at the end (hopefully less than ~2h)

For a piece of content with objectives harder to define, textbook-like or without a tangible result, maybe you should consider splitting it in several pieces, each with a clearly defined, hands-on, goal, that can be prerequisites to each other. 

With that said, we do have some pieces like that published, and we will keep links in the middle of the content when they bring more value to the reader's experience than they would in prerequisites or as an opening on wider topics at the end.

I also want to say that this tutorial is a high quality contribution, thank you!

Reply:
Yes I can appreciate that point. An up-front section before starting the tutorial, coupled with the existing section at the end, mentioning that understanding of HTML would help, and that resources for learning CSS & JS exist on MDN, would largely satisfy my desire here. This way you can still provide the useful resources, and provide them before people might have abandoned the tutorial.